src/highlighters.cc

```diff
@@ -971,29 +971,43 @@ void show_whitespaces(HighlightContext context, DisplayBuffer& display_buffer, B
 
             auto begin = get_iterator(buffer, atom_it->begin());
             auto end = get_iterator(buffer, atom_it->end());
+            auto last_non_space = begin.coord();
+
+            if(only_trailing)
+            {
+              for (BufferIterator it = begin; it != end; )
+              {
+                  Codepoint cp = utf8::read_codepoint(it, end);
+                  if (cp == '\t' or cp == ' ' or cp == '\n' or cp == 0xA0)
+                    continue;
+                  last_non_space = it.coord();
+              }
+            }
+
             for (BufferIterator it = begin; it != end; )
             {
                 auto coord = it.coord();
                 Codepoint cp = utf8::read_codepoint(it, end);
                 if (cp == '\t' or cp == ' ' or cp == '\n' or cp == 0xA0)
                 {
+                    auto should_highlight = !only_trailing or it.coord() > last_non_space;
                     if (coord != begin.coord())
                         atom_it = ++line.split(atom_it, coord);
                     if (it != end)
                         atom_it = line.split(atom_it, it.coord());
 
-                    if (cp == '\t')
+                    if (cp == '\t' and should_highlight)
```

Comment:
Maybe wrap this whole `if - else if - else if` block in another `if (should_highlight)` to avoid duplicating this condition all the time